test/distributed/org/apache/cassandra/distributed/test/ReadCoordinationMetricsTest.java

```diff
@@ -0,0 +1,124 @@
+    {
+        return node.callOnInstance(() -> ReadCoordinationMetrics.nonreplicaRequests.getCount());
+    }
+
+    private static long countPreferredOtherReplicas(IInvokableInstance node)
+    {
+        return node.callOnInstance(() -> ReadCoordinationMetrics.preferredOtherReplicas.getCount());
+    }
+
+    /**
+     * Two nodes with RF=1 so half the data will be owned by each node and the coordinator for queries is not
+     * always a replica in the list of candidates in {@link AbstractReadExecutor#getReadExecutor} where
+     * {@link ReadCoordinationMetrics#nonreplicaRequests} will be incremented.
+     *
+     * @throws Throwable
+     */
+    @Test
+    public void testNonReplicaRequests() throws Throwable
+    {
+        try (Cluster cluster = init(Cluster.create(2), 1))
+        {
+            cluster.schemaChange(withKeyspace("CREATE TABLE %s.tbl (pk int, ck int, v int, PRIMARY KEY (pk, ck))"));
+            for (int i = 0; i < 100; i++)
+                cluster.coordinator(1).execute(withKeyspace("INSERT INTO %s.tbl (pk, ck, v) VALUES (?,?,?)"), ALL, i, i, i);
+
+            long nonReplicaRequests1 = countNonreplicaRequests(cluster.get(1));
+            long nonReplicaRequests2 = countNonreplicaRequests(cluster.get(2));
+
+            for (int i = 0; i < 100; i++)
+            {
+                // When the coordinator is not a candidate replica, which will be half the time due to RF=1,
+                // the non-replica count metric will be incremented.
+                cluster.coordinator(1).execute(withKeyspace("SELECT * FROM %s.tbl WHERE pk = ? and ck = ?"), ALL, i, i);
+                cluster.coordinator(2).execute(withKeyspace("SELECT * FROM %s.tbl WHERE pk = ? and ck = ?"), ALL, i, i);
+            }
+
+            nonReplicaRequests1 = countNonreplicaRequests(cluster.get(1)) - nonReplicaRequests1;
+            nonReplicaRequests2 = countNonreplicaRequests(cluster.get(2)) - nonReplicaRequests2;
+            Assert.assertEquals(100, nonReplicaRequests1 + nonReplicaRequests2);
+        }
+    }
+
+    /**
+     * Two nodes with RF=2 so that both nodes are replicas for all data; this ensures that the coordinator node for
+     * queries will be a replica in the list of candidates in {@link AbstractReadExecutor#getReadExecutor}.
+     * <p>
+     * When the candidates collection is created, the sort order is changed so that the coordinator node is last. Then,
+     * using with CL=1 in the query, the resulting contacts collection will not contain the coordinator node, causing
+     * {@link ReadCoordinationMetrics#preferredOtherReplicas} to be incremented.
+     *
+     * @throws Throwable
+     */
+    @Test
+    public void testPreferredOtherReplicas() throws Throwable
+    {
+        try (Cluster cluster = init(Cluster.create(2), 2))
+        {
+            cluster.schemaChange(withKeyspace("CREATE TABLE %s.tbl (pk int, ck int, v int, PRIMARY KEY (pk, ck))"));
+            for (int i = 0; i < 100; i++)
```

Comment:
Since the magic number `100` is number of rows inserted, queried and expected, will it be good to put it into a constant?